liquidprompt

```diff
@@ -150,6 +150,167 @@ case "$(uname)" in
+        # We can get the arch along, with a single subshell.
+        if ! uname_sm="$(uname -s -m)"; then
+            return 1
+        fi
+        IFS=' ' read -r kernel lp_os_arch <<<"$uname_sm"
+    else # In any case, we need to know the kernel.
+        kernel="$(uname -s)"
+    fi
+
+    if (( LP_ENABLE_OS_KERNEL
+       || LP_ENABLE_OS_FAMILY
+       || LP_ENABLE_OS_DISTRIB
+       || LP_ENABLE_OS_VERSION )); then
+
+        case "$kernel" in
+            FreeBSD)
+                if (( LP_ENABLE_OS_FAMILY )); then lp_os_family=BSD; fi
+                if (( LP_ENABLE_OS_KERNEL )); then lp_os_kernel=FreeBSD; fi
+                ;;
+            DragonFly)
+                if (( LP_ENABLE_OS_FAMILY )); then lp_os_family=BSD; fi
+                if (( LP_ENABLE_OS_KERNEL )); then lp_os_kernel=DragonFly; fi
+                ;;
+            OpenBSD)
+                if (( LP_ENABLE_OS_FAMILY )); then lp_os_family=BSD; fi
+                if (( LP_ENABLE_OS_KERNEL )); then lp_os_kernel=OpenBSD; fi
+                ;;
+            Darwin)
+                if (( LP_ENABLE_OS_FAMILY )); then lp_os_family=BSD; fi
+                if (( LP_ENABLE_OS_KERNEL )); then lp_os_kernel=Darwin; fi
+                ;;
+            SunOS)
+                if (( LP_ENABLE_OS_FAMILY )); then lp_os_family=UNIX; fi
+                if (( LP_ENABLE_OS_KERNEL )); then lp_os_kernel=SunOS; fi
+                ;;
+            CYGWIN*)
+                if (( LP_ENABLE_OS_FAMILY )); then lp_os_family=Windows; fi
+                if (( LP_ENABLE_OS_KERNEL )); then lp_os_kernel=Cygwin; fi
+                ;;
+            MSYS*)
+                if (( LP_ENABLE_OS_FAMILY )); then lp_os_family=Windows; fi
+                if (( LP_ENABLE_OS_KERNEL )); then lp_os_kernel=MSYS; fi
+                ;;
+            MINGW*)
+                if (( LP_ENABLE_OS_FAMILY )); then lp_os_family=Windows; fi
+                if (( LP_ENABLE_OS_KERNEL )); then lp_os_kernel=MinGW; fi
+                ;;
+            *) # Assume Linux
+                if (( LP_ENABLE_OS_KERNEL )); then lp_os_kernel=Linux; fi
+                if (( LP_ENABLE_OS_FAMILY )); then
+                    if [[ "$kernel" == "Linux" ]]; then
+                        osfamily="$(uname -o)"
+                        if [[ "$osfamily" == "Android" ]]; then
+                            lp_os_family=Android
+                            return
+                        fi # Android
+                    fi # Linux
+                    lp_os_family=GNU
+                fi
+                # TODO parse HOME_URL and add an hyperlink.
+                if (( LP_ENABLE_OS_DISTRIB && LP_ENABLE_OS_VERSION )); then
+                    if _lp_grep_fields "/etc/os-release" "=" "ID" "VERSION_CODENAME"; then
+                        lp_os_distrib=${lp_grep_fields[_LP_FIRST_INDEX+0]-}
+                        lp_os_version=${lp_grep_fields[_LP_FIRST_INDEX+1]-}
+                    else
+                        return "$?"
```

Comment:
Why does this need to be an error? Sure, the docs as written back this up:
> Returns ``1`` if some expected information was missing.

But why does this need to be the case? Usually, we do a best effort on data sources, returning as much information as we can get.

And even if that did make sense, why `return "$?"`? The docs suggest this case should be `return 1`.

I say the only case where returning 1 makes sense is if we have no data at all to return.

Reply:
Makes sense. But then I think we need to avoid Linux being the fallback, and just let `lp_os_kernel=""`.

This is then one reason to separate the OS detection for LP_OS and for the data display.

Reply:
I agree. And it looks like you have done that change already.